crates/bindings/src/lib.rs

```diff
@@ -45,13 +42,610 @@ pub use timestamp::Timestamp;
+/// use spacetimedb::reducer;
+/// use log::info;
+///
+/// #[reducer]
+/// pub fn hello_world(context: &ReducerContext) {
+///     info!("Hello, World!");
+/// }
+///
+/// #[reducer]
+/// pub fn add_person(context: &ReducerContext, name: String, age: u16) {
+///     // add a "person" to the database.
+/// }
+///
+/// #[derive(SpacetimeType)]
+/// struct Coordinates {
+///     x: f32,
+///     y: f32,
+/// }
+///
+/// #[derive(Debug)]
+/// enum AddPlaceError {
+///     InvalidCoordinates(Coordinates),
+///     InvalidName(String),
+/// }
+///
+/// #[reducer]
+/// pub fn add_place(
+///     context: &ReducerContext,
+///     name: String,
+///     x: f32,
+///     y: f32,
+///     area: f32,
+/// ) -> Result<(), AddPlaceError> {
+///     // add a "place" to the database.
+/// }
+/// ```
+///
+/// Reducers may fail by returning a [`Result::Err`](std::result::Result) or by [panicking](std::panic!).
+/// Such a failure will be printed to the module logs and abort the active database transaction.
+/// Any changes to the database will be rolled back.
+///
+/// Reducers are limited in their ability to interact with the outside world.
+/// They do not directly return data aside from errors, and have no access to any
+/// network or filesystem interfaces.
+/// Calling methods from [`std::io`], [`std::net`], or [`std::fs`]
+/// inside a reducer will result in runtime errors.
+///
+/// Reducers can communicate information to the outside world in two ways:
+/// - They can modify tables in the database.
+///   See the `#[table]`(#table) macro documentation for information on how to declare and use tables.
+/// - They can call logging macros from the [`log`] crate.
+///   This writes to a private debug log attached to the database.
+///   Run `spacetime logs <DATABASE_IDENTITY>` to browse these.
+///
+/// Reducers are permitted to call other reducers, simply by passing their `ReducerContext` as the first argument.
+/// This is a regular function call, and does not involve any network communication. The callee will run within the
+/// caller's transaction, and any changes made by the callee will be committed or rolled back with the caller.
+///
+/// # Lifecycle Reducers
+///
+/// You can specify special lifecycle reducers that are run at set points in
+/// the module's lifecycle. You can have one of each per module.
+///
+/// These reducers cannot be called manually
+/// and may not have any parameters except for `ReducerContext`.
+///
+/// ### `#[spacetimedb::reducer(init)]`
+///
+/// This reducer is run the first time a module is published
+/// and any time the database is cleared.
+///
+/// If an error occurs when initializing, the module will not be published.
+///
+/// ### `#[spacetimedb::reducer(client_connected)]`
+///
+/// This reducer is run when a client connects to the SpacetimeDB module.
+/// Their identity can be found in the sender value of the `ReducerContext`.
+///
+/// If an error occurs in the reducer, the client will be disconnected.
```

Comment:
I don't think we've implemented this behavior. CC @jsdt .

Reply:
I didn't implement it, but from looking at the code, it looks like that is true. That's the first thing we do when trying to spawn a client connection [here](https://github.com/clockworklabs/SpacetimeDB/blob/a54ea3a391c2828c6fa49bb905312e0154be6b40/crates/core/src/client/client_connection.rs#L184), and we give up if it fails.